apps/docs/content/guides/auth/custom-claims-and-role-based-access-control-rbac.mdx

```diff
@@ -0,0 +1,259 @@
+
+```json
+"app_metadata": {
+  "plan": "TRIAL",
+  "user_level": 100,
+  "group_name": "Super Guild!",
+  "joined_on": "2022-05-20T14:28:18.217Z",
+  "group_manager": false,
+  "items": ["toothpick", "string", "ring"],
+}
+```
+
+While there are many ways to implement `custom claims` we recommend utilizing the [custom access token auth hook](/guides/auth/auth-hooks#hook-custom-access-token), which runs before a token is issued and allows you to add additional claims to the user's JWT.
+
+This guide uses the [Slack Clone example](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone) to demonstrate how to add a `user_role` claim and use it in your [Row Level Security (RLS) policies](/guides/auth/row-level-security).
+
+## Create a table to track user roles and permissions
+
+In this example we're implementing two user roles with specific permissions:
+
+- `moderator`: A moderator can delete all messages but not channels.
+- `admin`: An admin can delete all messages and channels.
+
+```sql supabase/migrations/init.sql
+-- Custom types
+create type public.app_permission as enum ('channels.delete', 'messages.delete');
+create type public.app_role as enum ('admin', 'moderator');
+
+-- USER ROLES
+create table public.user_roles (
+  id        bigint generated by default as identity primary key,
+  user_id   uuid references public.users on delete cascade not null,
+  role      app_role not null,
+  unique (user_id, role)
+);
+comment on table public.user_roles is 'Application roles for each user.';
+
+-- ROLE PERMISSIONS
+create table public.role_permissions (
+  id           bigint generated by default as identity primary key,
+  role         app_role not null,
+  permission   app_permission not null,
+  unique (role, permission)
+);
+comment on table public.role_permissions is 'Application permissions for each role.';
+```
+
+<Admonition type="note">
+  
+  For the [full schema](https://github.com/supabase/supabase/blob/master/examples/slack-clone/nextjs-slack-clone/README.md) please refer to the example application on [Github](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone).
+
+</Admonition>
+
+You can now manage your roles and permissions in SQL, e.g. to add the mentioned roles and permissions from aboove you can say:
+
+```sql supabase/seed.sql
+insert into public.role_permissions (role, permission)
+values
+  ('admin', 'channels.delete'),
+  ('admin', 'messages.delete'),
+  ('moderator', 'messages.delete');
+```
+
+## Create auth hook to apply user role
+
+The [custom access token auth hook](/guides/auth/auth-hooks#hook-custom-access-token) runs before a token is issued allowing you to add additional claims to the user's JWT. You should add the custom claim to the user's `app_metadata` to make sure that the logged in user can't modify this value themselves!
```

Comment:
```suggestion
The [Custom Access Token Auth Hook](/guides/auth/auth-hooks#hook-custom-access-token) runs before a token is issued. You can use it edit the JWT. You should add your custom claim to the user's `app_metadata` to make sure that the logged-in user can't modify this value themselves!
```

Should it be `raw_app_metadata`? I have only vague memories of seeing an issue about this so I could be wrong